RestaurantReservation.API/Controllers/CustomersController.cs

```diff
@@ -0,0 +1,162 @@
+﻿using AutoMapper;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.JsonPatch;
+using Microsoft.AspNetCore.Mvc;
+using RestaurantReservation.API.Models.Customers;
+using RestaurantReservation.Db.Interfaces;
+using RestaurantReservation.Db.Models.Entities;
+using System.Text.Json;
+
+namespace RestaurantReservation.API.Controllers
+{
+    [Route("api/Customers")]
+    [ApiController]
+    [Authorize]
+    public class CustomersController : ControllerBase
+    {
+        private readonly ICustomerRepository _customerRepository;
+        private readonly IMapper _mapper;
+        private const int MaxPageSize = 5;
+
+        public CustomersController(ICustomerRepository customerRepository, IMapper Mapper)
+        {
+            _customerRepository = customerRepository;
+            _mapper = Mapper;
+        }
+
+        /// <summary>
+        /// Gets all customers with pagination.
+        /// </summary>
+        /// <param name="pageNumber">The number of the page to retrieve.</param>
+        /// <param name="pageSize">The number of items per page.</param>
+        /// <returns>A list of customer DTOs.</returns>
+        [HttpGet]
+        [ProducesResponseType(StatusCodes.Status200OK)]
+        [ProducesResponseType(StatusCodes.Status204NoContent)]
+        [ProducesResponseType(StatusCodes.Status400BadRequest)]
+        public async Task<ActionResult<IEnumerable<CustomerDto>>> GetAllCustomers(int pageNumber = 1, int pageSize = 10)
+        {
+            if (pageNumber < 1 || pageSize < 1)
+                return BadRequest($"'{nameof(pageNumber)}' and '{nameof(pageSize)}' must be greater than 0.");
+
+            pageSize = Math.Min(pageSize, MaxPageSize);
+
+            var (customers, paginationMetadata) = await _customerRepository.GetAllAsync(_ => true, pageNumber, pageSize);
+
+            Response.Headers.Add("X-Pagination", JsonSerializer.Serialize(paginationMetadata));
+
+            return Ok(_mapper.Map<IEnumerable<CustomerDto>>(customers));
+        }
+
+        /// <summary>
+        /// Retrieves a customer by their ID.
+        /// </summary>
+        /// <param name="id">The ID of the customer to retrieve.</param>
+        /// <returns>A customer DTO if found; otherwise, a 404 Not Found response.</returns>
+        [HttpGet("{id}", Name = "GetCustomer")]
+        [ProducesResponseType(StatusCodes.Status200OK)]
+        [ProducesResponseType(StatusCodes.Status404NotFound)]
+        public async Task<ActionResult<CustomerDto>> GetCustomer(int id)
+        {
+            var customer = await _customerRepository.GetByIdAsync(id);
+            if (customer == null)
```

Comment:
because you have the nullable annotation enabled, you need to mark the object as a nullable value before checking if it's null. the way it's implemented it will fire a different exception if it was null